File: plugins/ruby/crepuscularity_plugin.rb
require "json"
require "open3"
require "cgi"

module CrepuscularityPlugin
  ViewIr = Struct.new(:version, :root, keyword_init: true)

  BIND_BLOCKLIST = ["baseDir", "_"].freeze # ponytail: block security-sensitive keys only

  class ViewSession
    attr_reader :path, :context

    def initialize(path, context = {})
      @path = path
      @context = context.dup
      @handlers = {}
    end

    def on(handler, callback = nil, &block)
      @handlers[handler] = callback || block
      self
    end

    def render_ir
      CrepuscularityPlugin.render_ir(@path, @context)
    end

    def render_html
      render_ir.root.map { |node| CrepuscularityPlugin.render_node(node) }.join
    end

    def dispatch(event)
      payload = event.is_a?(String) ? { "handler" => event } : event
      handler = payload.fetch("handler", "").to_s
      if handler.start_with?("bind:")
        key, value = handler.delete_prefix("bind:").split(":", 2)
        @context[key] = value unless value.nil? || BIND_BLOCKLIST.include?(key)
      end
      @handlers[handler]&.call(payload, self)
      render_ir
    end
  end

  def self.render_ir(path, context = nil)
    # resolve symlinks and absolute paths securely
    resolved_path = File.realpath(path)
    # ensure it strictly resides within the current working directory boundary
    base_dir = File.realpath(Dir.pwd)
    unless resolved_path.start_with?(base_dir + File::SEPARATOR) || resolved_path == base_dir
      raise ArgumentError, "Invalid path: must be within current directory"
    end

    bin = ENV.fetch("CREPUS_BIN", "crepus")
    if context
      payload = JSON.generate({
        "template" => File.read(path),
        "context" => context,
        "baseDir" => File.dirname(path)
      })
      stdout, stderr, status = Open3.capture3(bin, "native", "ir", "--stdin-json", stdin_data: payload)
    else
      stdout, stderr, status = Open3.capture3(bin, "native", "ir", path)
    end
    raise stderr unless status.success?
    data = JSON.parse(stdout)
    ViewIr.new(version: data.fetch("version"), root: data.fetch("root"))
  end

  def self.render_html(path, context = nil)
    render_ir(path, context).root.map { |node| render_node(node) }.join
  end

  def self.render_node(node)
    case node["kind"]
    when "text"
      CGI.escapeHTML(node.fetch("content", "").to_s)
    when "stack", "scroll"
      children = node.fetch("children", []).map { |child| render_node(child) }.join
      %(<div data-crepus-kind="#{CGI.escapeHTML(node["kind"].to_s)}" data-axis="#{CGI.escapeHTML(node.fetch("axis", "column").to_s)}">#{children}</div>)
    when "button"
      label = CGI.escapeHTML(node.fetch("label", "").to_s)
      node["onClick"] ? %(<button data-onclick="#{CGI.escapeHTML(node["onClick"].to_s)}">#{label}</button>) : "<button>#{label}</button>"
    when "image"
      %(<img src="#{CGI.escapeHTML(node.fetch("src", "").to_s)}" alt="#{CGI.escapeHTML(node.fetch("alt", "").to_s)}">)
    when "slotRotate"
      %(<span data-crepus-kind="slotRotate">#{CGI.escapeHTML(node.fetch("phrases", [""]).first.to_s)}</span>)
    when "input"
      bind = CGI.escapeHTML(node.fetch("bind", "").to_s)
      placeholder = CGI.escapeHTML(node.fetch("placeholder", "").to_s)
      node["multiline"] ? %(<textarea data-bind="#{bind}" placeholder="#{placeholder}"></textarea>) : %(<input data-bind="#{bind}" placeholder="#{placeholder}">)
    when "picker"
      options = node.fetch("options", []).map do |option|
        %(<option value="#{CGI.escapeHTML(option.fetch("value", "").to_s)}">#{CGI.escapeHTML(option.fetch("label", "").to_s)}</option>)
      end.join
      %(<select data-bind="#{CGI.escapeHTML(node.fetch("bind", "").to_s)}">#{options}</select>)
    else
      ""
    end
  end
end

File: plugins/python/crepuscularity_plugin.py
from __future__ import annotations

import json
import os
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Callable
from html import escape


@dataclass
class ViewIr:
    version: int
    root: list[dict[str, Any]]


EventPayload = dict[str, Any]
EventHandler = Callable[[EventPayload, "ViewSession"], None]


class ViewSession:
    def __init__(self, path: str | Path, context: dict[str, Any] | None = None, allowed_dir: str | Path | None = None) -> None:
        self.path = Path(path)
        self.context = dict(context or {})
        self.allowed_dir = allowed_dir
        self._handlers: dict[str, EventHandler] = {}

    def on(self, handler: str, callback: EventHandler) -> "ViewSession":
        self._handlers[handler] = callback
        return self

    def render_ir(self) -> ViewIr:
        return render_ir(self.path, self.context, self.allowed_dir)

    def render_html(self) -> str:
        return "".join(_render_node(node) for node in self.render_ir().root)

    def dispatch(self, event: str | EventPayload) -> ViewIr:
        payload = {"handler": event} if isinstance(event, str) else event
        handler = str(payload.get("handler", ""))
        if handler.startswith("bind:"):
            parts = handler.removeprefix("bind:").split(":", 1)
            # ponytail: block security-sensitive keys only
            if len(parts) == 2 and parts[0] not in _BIND_BLOCKLIST:
                self.context[parts[0]] = parts[1]
        callback = self._handlers.get(handler)
        if callback is not None:
            callback(payload, self)
        return self.render_ir()


def _crepus_bin() -> str:
    bin_path = os.environ.get("CREPUS_BIN", "crepus")
    name = os.path.basename(bin_path)
    if name not in ("crepus", "crepus.exe"):
        raise ValueError(f"Invalid CREPUS_BIN: binary name must be 'crepus' or 'crepus.exe', got {name!r}")
    if bin_path != name and not os.path.isabs(bin_path):
        raise ValueError("Invalid CREPUS_BIN: must be an absolute path or a simple binary name")
    return bin_path


_BIND_BLOCKLIST = frozenset({"baseDir", "_"})  # ponytail: block security-sensitive keys only


def render_ir(path: str | Path, context: dict[str, Any] | None = None, allowed_dir: str | Path | None = None) -> ViewIr:
    resolved_path = Path(path).resolve()
    resolved_allowed = Path(allowed_dir).resolve() if allowed_dir is not None else Path.cwd().resolve()
    if not resolved_path.is_relative_to(resolved_allowed):
        raise ValueError("Path traversal detected")

    args = [_crepus_bin(), "native", "ir", str(path)]
    input_data = None
    if context is not None:
        source = resolved_path.read_text()
        payload = {"template": source, "context": context, "baseDir": str(Path(path).parent)}
        args = [_crepus_bin(), "native", "ir", "--stdin-json"]
        input_data = json.dumps(payload)
    proc = subprocess.run(args, input=input_data, text=True, capture_output=True, check=False)
    if proc.returncode != 0:
        raise RuntimeError(proc.stderr.strip())
    data = json.loads(proc.stdout)
    return ViewIr(version=data["version"], root=data["root"])


def render_html(path: str | Path, context: dict[str, Any] | None = None, allowed_dir: str | Path | None = None) -> str:
    return "".join(_render_node(node) for node in render_ir(path, context, allowed_dir).root)


def _render_node(node: dict[str, Any]) -> str:
    kind = node.get("kind")
    if kind == "text":
        return escape(str(node.get("content", "")))
    if kind == "stack":
        axis = escape(str(node.get("axis", "column")))
        children = "".join(_render_node(child) for child in node.get("children", []))
        return f'<div data-crepus-kind="stack" data-axis="{axis}">{children}</div>'
    if kind == "scroll":
        axis = escape(str(node.get("axis", "column")))
        children = "".join(_render_node(child) for child in node.get("children", []))
        return f'<div data-crepus-kind="scroll" data-axis="{axis}">{children}</div>'
    if kind == "button":
        label = escape(str(node.get("label", "")))
        on_click = escape(str(node.get("onClick", "")))
        attr = f' data-onclick="{on_click}"' if on_click else ""
        return f"<button{attr}>{label}</button>"
    if kind == "image":
        src = escape(str(node.get("src", "")), quote=True)
        alt = escape(str(node.get("alt", "")), quote=True)
        return f'<img src="{src}" alt="{alt}">'
    if kind == "slotRotate":
        phrase = escape(str((node.get("phrases") or [""])[0]))
        return f'<span data-crepus-kind="slotRotate">{phrase}</span>'
    if kind == "input":
        placeholder = escape(str(node.get("placeholder", "")), quote=True)
        bind = escape(str(node.get("bind", "")), quote=True)
        if node.get("multiline"):
            return f'<textarea data-bind="{bind}" placeholder="{placeholder}"></textarea>'
        return f'<input data-bind="{bind}" placeholder="{placeholder}">'
    if kind == "picker":
        bind = escape(str(node.get("bind", "")), quote=True)
        options = "".join(
            f'<option value="{escape(str(opt.get("value", "")), quote=True)}">{escape(str(opt.get("label", "")))}</option>'
            for opt in node.get("options", [])
        )
        return f'<select data-bind="{bind}">{options}</select>'
    return ""
